File: algoritmos/orden/fun_existente/orden.cpp
#include <algorithm>
#include <fstream>
#include <iostream>

int main() {
    //std::vector<int> vec = {3, 1, 4, 1, 5, 9, 2, 6, 5, 3};

    // archivo original
    std::ifstream file("../random_list.bin", std::ios::binary);
    if (!file) {
        std::cerr << "Error al abrir" << std::endl;
        return 1;
    }

    file.seekg(0, std::ios::end);
    long file_size = file.tellg();
    file.seekg(0, std::ios::beg);

    int num = file_size / sizeof(int);
    int *array_int = new int[num];

    file.read(reinterpret_cast<char*>(array_int), file_size);

    file.close();

    /*
    // txt para comprobar
    std::ofstream txt_file("random_list_cpp.txt");
    if (!txt_file) {
        std::cerr << "Error al crear el archivo de texto" << std::endl;
        delete[] array_int;
        return 1;
    }

    for (int i = 0; i < num; i++) {
        txt_file << array_int[i] << std::endl;
    }

    txt_file.close();
    // txt generado
    */

    std::sort(array_int, array_int + num);

    // txt para comprobar
    std::ofstream txt_file_sorted("sorted_list_cpp.txt");
    if (!txt_file_sorted) {
        std::cerr << "Error al crear el archivo de texto" << std::endl;
        delete[] array_int;
        return 1;
    }

    for (int i = 0; i < num; i++) {
        txt_file_sorted << array_int[i] << std::endl;
    }

    txt_file_sorted.close();
    // txt generado

    delete[] array_int;

    return 0;
}

File: algoritmos/orden/fun_existente/orden.java
import java.io.BufferedWriter;
import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;

public class orden {

    static int[] readRandomBin() {
        try (DataInputStream file = new DataInputStream(new FileInputStream("../random_list.bin"))) {

            long fileSize = file.available();
            int num = (int) (fileSize / Integer.BYTES);

            int[] array_int = new int[num];

            // lee el archivo binario en formato bytes
            byte[] array_byte = new byte[(int) fileSize];
            file.read(array_byte);

            // convierte al orden de bytes nativo
            ByteBuffer buffer = ByteBuffer.wrap(array_byte);
            buffer.order(ByteOrder.nativeOrder());

            for (int i = 0; i < num; ++i) {
                array_int[i] = buffer.getInt();
            }

            return array_int;
            
        } catch (IOException e) {
            System.err.println("Error al leer el archivo binario: " + e.getMessage());
            return null;
        }
    }

    static void createRandomFile(int[] array, int num) {
        try (BufferedWriter txtFile = new BufferedWriter(new FileWriter("random_list_java.txt"))) {
            for (int i = 0; i < num; ++i) {
                txtFile.write(Integer.toString(array[i]) + "\n");
            }
        } catch (IOException e) {
            System.err.println("Error al escribir en el archivo de texto: " + e.getMessage());
            return;
        }
    }

    static void createSortedFile(int[] array, int num) {
        try (BufferedWriter txtFile = new BufferedWriter(new FileWriter("sorted_list_java.txt"))) {
            for (int i = 0; i < num; ++i) {
                txtFile.write(Integer.toString(array[i]) + "\n");
            }
        } catch (IOException e) {
            System.err.println("Error al escribir en el archivo de texto: " + e.getMessage());
            return;
        }
    }

    public static void main(String[] args) {
        int[] arr = readRandomBin();
        int n = arr.length;

        /*
        createRandomFile(arr, n);
        */
        
        Arrays.sort(arr);

        createSortedFile(arr, n);
    }
}
